### ISSUE TITLE
[ReactNativeStore] IllegalStateException: site.wpApiRestUrl must not be null

### ISSUE THREAD

**Issue by oguzkocer** [wordpress-mobile/WordPress-FluxC-Android#1579]
**Sentry Url**: https://sentry.io/share/issue/fac21421243d44ee9ddb6127e35d7f12/
**User Count**: 10
**Count**: 19
**First Release**: org.wordpress.android-14.7
**First Seen**: 2020-05-08T16:46:59Z
**Last Seen**: 2020-05-11T01:26:50Z
**24 Hours**: 2
**30 Days**: 19

```
IllegalStateException: site.wpApiRestUrl must not be null
     at org.wordpress.android.fluxc.store.ReactNativeStore.executeWPAPIRequest(ReactNativeStore.kt:93)
     at org.wordpress.android.fluxc.store.ReactNativeStore.executeWPAPIRequest(ReactNativeStore.kt:79)
     at org.wordpress.android.fluxc.store.ReactNativeStore$executeRequest$2.invokeSuspend(ReactNativeStore.kt:62)
     at org.wordpress.android.fluxc.store.ReactNativeStore$executeRequest$2.invoke(null:10)
     at kotlinx.coroutines.intrinsics.UndispatchedKt.startUndispatchedOrReturn(Undispatched.kt:91)
     at kotlinx.coroutines.BuildersKt__Builders_commonKt.withContext(Builders.common.kt:154)
     at kotlinx.coroutines.BuildersKt.withContext(null:1)
     at org.wordpress.android.fluxc.tools.CoroutineEngine.withDefaultContext(CoroutineEngine.kt:23)
     at org.wordpress.android.fluxc.store.ReactNativeStore.executeRequest(ReactNativeStore.kt:58)
     at org.wordpress.android.ui.posts.reactnative.ReactNativeRequestHandler$performGetRequest$1.invokeSuspend(ReactNativeRequestHandler.kt:33)
...
(6 additional frame(s) were not displayed)
```

**Comment by mchowning** [comment]
Took a look at this and I believe the exception is getting thrown because `site.wpApiRestUrl` is `null` on the last line of [this code](https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/1.6.13/fluxc/src/main/java/org/wordpress/android/fluxc/store/ReactNativeStore.kt#L87-L93):

```
val usingSavedRestUrl = site.wpApiRestUrl != null
if (!usingSavedRestUrl) {
    site.wpApiRestUrl = discoveryWPAPIRestClient.discoverWPAPIBaseURL(site.url) // discover rest api endpoint
            ?: slashJoin(site.url, "wp-json/") // fallback to ".../wp-json/" default if discovery fails
    persistSiteSafely(site)
}
val fullRestUrl = slashJoin(site.wpApiRestUrl, path)
```

It seems like the only way that could happen is if `persistSiteSafely(site)` is mutating the site object. I don't have time to look into this more now, but I'll try to get back to it later.

cc: @malinajirka

**Comment by malinajirka** [comment]
It could theoretically be caused by a race condition - if a different thread modified the `site` object. 😕 ❓

**Comment by mchowning** [comment]
> It could theoretically be caused by a race condition - if a different thread modified the site object. 😕 ❓

You know, that might be it. There are frequently multiple API calls going on at the same time, and [one path in the request flow](https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/1.6.13/fluxc/src/main/java/org/wordpress/android/fluxc/store/ReactNativeStore.kt#L126) does set `site.wpApiRestUrl` to be `null`.

Thanks @malinajirka !

**Comment by fluiddot** [comment]
Heads up that I linked in the previous comment a new Sentry issue, it's almost identical to the one referenced in the description except that the cause is `NullPointerException` instead of `IllegalStateException `.

**Error description:**
```
NullPointerException
site.wpApiRestUrl must not be null
```

**Comment by AliSoftware** [comment]
Heads up that I linked two more Sentry reports ☝️  related to that same issue above. One is new (at least wasn't triaged/linked to a GitHub issue before today) and the other was previously linked to the WPAndroid issue so I replaced the link to point to this issue instead.

**Comment by antonis** [comment]
People with older version (18.8) of the app still experience this :(
Ref: https://a8c.sentry.io/share/issue/79f9cfda0f0841a2a091519b29c9543e/

**Comment by irfano** [comment]
> People with older version (18.8) of the app still experience this :( Ref: https://a8c.sentry.io/share/issue/79f9cfda0f0841a2a091519b29c9543e/

This issue was resolved in version 20.1, and all events from Sentry are before `20.1`. I'm marking this issue as resolved in Sentry by specifying the version.
